sodetlib/operations/bias_dets.py

```diff
@@ -201,3 +201,408 @@ def bias_to_rfrac(S, cfg, rfrac=0.5, bias_groups=None, iva=None,
+            fig.savefig(fname)
+            S.pub.register_file(fname, 'biasstep_rfrac', format='png', plot=True)
+            if not show_plots:
+                plt.close(fig)
+        
+        ## add in check if detectors are normal now
+        percentage_rn_0 = bsa_0.R0/bsa_0.R_n_IV
+        drop_from_normal = False
+        for bl in bg_overbias_needed:
+            mask_bg = np.where(bsa_0.bgmap==bl)[0]
+            per_bl_percentage_rn_0 = percentage_rn_0[mask_bg]
+            ## mask for normal
+            mask_normal = np.where((per_bl_percentage_rn_0 > 0.9) | (per_bl_percentage_rn_0 < -0.9))
+            ## if more than half of the detectors are normal
+            if len(mask_normal[0]) > 0.5*len(mask_bg):
+                bg_detectors_normal.append(bl)   
+                drop_from_normal = True
+
+## drop normal detector into transition first
+    while drop_from_normal:
+        S.log(f"some biaslines are still in normal state: {bg_detectors_normal}")
+        percentage_rn_0 = bsa_0.R0/bsa_0.R_n_IV
+        S.log("droping detectors from normal to transition")
+        previous_dc_biases_dfn = S.get_tes_bias_bipolar_array()
+
+        S.log(f"current tes bias voltages are: {previous_dc_biases_dfn}")
+        for bl in bg_detectors_normal:
+            mask_bg = np.where(bsa_0.bgmap==bl)[0]
+            v_sc = iva.v_bias[iva.idxs[[mask_bg], 0]]
+            v_norm = iva.v_bias[iva.idxs[[mask_bg], 1]]
+            v_spread = np.nanmedian(v_norm) - np.nanmedian(v_sc)
+            if previous_dc_biases_dfn[bl] > v_spread:
+                previous_dc_biases_dfn[bl] =  previous_dc_biases_dfn[bl] - 0.5 * v_spread
+            else:
+                previous_dc_biases_dfn[bl] =  previous_dc_biases_dfn[bl] - 0.3 * v_spread 
+                bg_detectors_normal.remove(bl)
+                S.log(f'biasline {bl} is approching 0 voltage, a confirmation new IV curve is recommended')
+        
+        S.set_tes_bias_bipolar_array(previous_dc_biases_dfn)
+        S.log(f"applying: {previous_dc_biases_dfn}")
+        S.log(f"waiting 10s for dissipation of UFM local heating")
+        time.sleep(10)
+        bsa_0 = bias_steps.take_bias_steps(S, cfg)
+        percentage_rn_0 = bsa_0.R0/bsa_0.R_n_IV
+        drop_from_normal = False
+        for bl in bg_detectors_normal:
+            mask_bg = np.where(bsa_0.bgmap==bl)[0]
+            per_bl_percentage_rn_0 = percentage_rn_0[mask_bg]
+            ## mask for normal
+            mask_normal = np.where((per_bl_percentage_rn_0 > 0.9) | (per_bl_percentage_rn_0 < -0.9))
+            ## if more than half of the detectors are normal
+            if len(mask_normal[0]) > 0.5*len(mask_bg):
+                drop_from_normal = True
+            else:
+                bg_detectors_normal.remove(bl)
+        for bl in bg_detectors_normal:
+            if previous_dc_biases_dfn[bl] < 0.5:
+                bg_detectors_normal.remove(bl)
+
```

Comment:
Can you explain a bit about what this is doing? I don't quite understand how you're chosing the bias voltage step sizes, and also it's not clear to me from the code how many times this runs on average, but seems like it could be a lot of bias step measurements. Do you think this is mainly what's dominating the function's runtime?

Reply:
I think we might want to brainstorm other ways to drop into the transition. I think if we're taking small enough step sizes we may not need to take bias steps every time, since we're unlikely to induce tracking skips.

Reply:
this is looping down the bias voltages until a certain bias line is no longer in normal state and TES start to be in-transtion. The step size being the 50% of the transition region from the latest IV such that the bias_line won't be dropped into SC due to step being too large or the script takes too long due to the step being to small. Notice the latest IV doesn't necessary provides good estimation as the I-V won't be taken as frequently as the bias-step, so a larger step size here can be dangerous. 
in the field I think the function runtime is dominated by the bias-step analysis itself. This is because in the field the sky loading changes gradually and slowly. the longest runtime will be the wait time of biasing detectors out of SC state, but this shouldn't happen if the re-biasing is done frequently enough nor should dropping from normal state happen frequently.